File: 219.py
class Solution:
    def containsNearbyDuplicate(self, nums: list[int], k: int) -> bool:
        if k == 0:
            return False
        s = set()
        n = min(k, len(nums))
        for i in range(n):
            if nums[i] in s:
                return True
            s.add(nums[i])
        r = n
        l = 0
        while r < len(nums):
            if nums[r] in s:
                return True
            else:
                s.remove(nums[l])
                s.add(nums[r])
            r += 1
            l += 1
        return False

File: 219.cpp
#include <iostream>
#include <map>
#include <set>
#include <vector>
using namespace std;

class Solution {
public:
  bool containsNearbyDuplicate(vector<int> &nums, int k) {
    map<int, int> m;
    for (int i = 0; i < nums.size(); ++i) {
      if (m.find(nums[i]) != m.end() && i - m[nums[i]] <= k) {
        return true;
      }
      m[nums[i]] = i;
    }
    return false;
  }
};

int main() {
  Solution s;
  vector<int> t1_in = {1, 2, 3, 1, 2, 3};
  bool res = s.containsNearbyDuplicate(t1_in, 2);
  cout << res << endl;
}